physics/discrete_trajectory_body.hpp

```diff
@@ -33,7 +33,7 @@ DiscreteTrajectory<Frame>::DiscreteTrajectory()
 template<typename Frame>
 typename DiscreteTrajectory<Frame>::reference
 DiscreteTrajectory<Frame>::front() const {
-  return *begin();
+  return segments_->front().front();
```

Comment:
We don’t think this is correct: `segments_->front()` can be empty; the design doc is out of date, as all design docs are, but there are comments to that effect: https://github.com/mockingbirdnest/Principia/blob/d9e6442827e5006d2e700d1baf45ccef03f25a35/physics/discrete_trajectory.hpp#L99-L101 https://github.com/mockingbirdnest/Principia/blob/d9e6442827e5006d2e700d1baf45ccef03f25a35/physics/discrete_trajectory.hpp#L218-L223

Reply:
Arguably, the fact that it is incorrect—and the fact that it is a performance improvement—seems like a bug in the implementation of `front` as changed in #3207 ; we end up with a `front()` with semantics that differ from `*begin()`, which seems distasteful.

In any case, please add a test that would have caught this.

It might be possible to actually optimize things using `segment_by_left_endpoint_`.

Reply:
OK, having discussed this with @pleroy, I think we want the following:
- fix `front` on segments (to do `*begin` again; #3207 probably predates the existence of empty segments, so it would have been correct at the time, but `front` on a segment is very rarely used, so nobody should care about its performance anyway);
- add a test that would catch brokenness on `front` on either an empty segment or a trajectory starting with empty segments;
- use `segments_by_left_endpoint_` to write a faster `front` on trajectories (maybe you could even use it to write a faster begin, and keep `front` as `*begin`?).

Reply:
Update: I have a test for `front` on a trajectory starting with an empty segment, and a fast correct version of trajectory `front` using `segments_by_left_endpoint_`. I'll fix segment `front` and look into improving trajectory `begin`, as requested.

Reply:
I wasn't able to speed up `begin`.
Here are the benchmarks for the updated functions:
```
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_DiscreteTrajectoryFront             0.592 ns        0.592 ns   1000000000
BM_DiscreteTrajectorySegmentFront      0.526 ns        0.525 ns   1000000000
```
```
Benchmark                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
BM_DiscreteTrajectoryFront                       -0.7197         -0.7196             2             1             2             1
BM_DiscreteTrajectorySegmentFront                +0.6716         +0.6716             0             1             0             1
```

Reply:
> I wasn't able to speed up `begin`.

Since you return a `DiscreteTrajectoryIterator`, and thus need to go through the constructor of `DiscreteTrajectoryIterator`, I don’t think you can make it faster in the simple case (no empty segments at the beginning), which is what you are benchmarking.

However, in हरीश चंद्र, we can reasonably expect to have hundreds of empty segments at the beginning (for instance, an old LEO sat that did occasional corrections with RCS will have two segments—burn and coast—for every RCS firing; most of those will not have been reanimated unless you looked at its entire history all the way to launch), so I suspect this may well be worth it.

Reply:
> we can reasonably expect to have hundreds of empty segments at the beginning
I'll go ahead and add another benchmark then!

Reply:
Benchmarks for trajectories beginning with 100 empty segments:
Pre-optimizations (note that this is with the buggy `Segment::front` code):
```
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_DiscreteTrajectoryFrontEmpty               113 ns          113 ns      6195622
BM_DiscreteTrajectoryBeginEmpty              88.8 ns         88.8 ns      7838307
BM_DiscreteTrajectorySegmentFrontEmpty      0.315 ns        0.315 ns   1000000000
BM_DiscreteTrajectorySegmentBeginEmpty       88.6 ns         88.6 ns      7969942
```
Post-optimizations (including the change to `DiscreteTrajectory::begin`):
```
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_DiscreteTrajectoryFrontEmpty             0.592 ns        0.592 ns   1000000000
BM_DiscreteTrajectoryBeginEmpty              1.46 ns         1.46 ns    479777383
BM_DiscreteTrajectorySegmentFrontEmpty        110 ns          110 ns      6333409
BM_DiscreteTrajectorySegmentBeginEmpty       88.5 ns         88.4 ns      7939659
```
Delta:
```
Benchmark                                                Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryFrontEmpty                       -0.9948         -0.9948           113             1           113             1
BM_DiscreteTrajectoryBeginEmpty                       -0.9836         -0.9836            89             1            89             1
BM_DiscreteTrajectorySegmentFrontEmpty              +349.6114       +349.5768             0           110             0           110
BM_DiscreteTrajectorySegmentBeginEmpty                -0.0016         -0.0014            89            88            89            88
```
`Segment::front` sees a massive regression (because it was broken before), while `DiscreteTrajectory::front` and `DiscreteTrajectory::begin` see massive gains.
The change to `DiscreteTrajectory::begin` causes a mild regression for the non-empty-segment benchmark:
```
Benchmark                                    Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryBegin                +0.1961         +0.1960             1             1             1             1
```
It's still under 2 ns though so I think it's ok.